experimental/robust_segvit/metrics_multihost.py

```diff
@@ -0,0 +1,148 @@
+"""Calculate ood metrics across hosts."""
+from typing import Any, Optional, Dict
+
+import jax
+import jax.numpy as jnp
+import tensorflow as tf
+from jax.experimental import multihost_utils
+from ood_metrics import get_ood_score
+from ood_metrics import get_score
+
+
+def keras_auc_to_arrays(keras_auc_object):
+  """Pull out arrays from keras roc object."""
+  # The thresholds used are determinisitc, so we need not store them.
+  tp = jnp.asarray(keras_auc_object.true_positives)
+  fp = jnp.asarray(keras_auc_object.false_positives)
+  tn = jnp.asarray(keras_auc_object.true_negatives)
+  fn = jnp.asarray(keras_auc_object.false_negatives)
+  return tp, fp, tn, fn
+
+
+def arrays_to_keras_auc(tp, fp, tn, fn, keras_auc_object):
+  """Assign confusion matrix arrays to a keras_auc_object."""
+  keras_auc_object.true_positives.assign(tp)
+  keras_auc_object.false_positives.assign(fp)
+  keras_auc_object.true_negatives.assign(tn)
+  keras_auc_object.false_negatives.assign(fn)
+  return keras_auc_object
+
+
+def combine_states(all_auc_states, num_thresholds=200):
+  # jax can take in trees of arrays, tuple is considered a tree so we can
+  # unpack it here.
+  # each array here has dimensions #host x shape
+
+  all_tp, all_fp, all_tn, all_fn = all_auc_states
+
+  assert all_tp.shape == (jax.process_count(), num_thresholds)
+  assert all_fp.shape == (jax.process_count(), num_thresholds)
+  assert all_tn.shape == (jax.process_count(), num_thresholds)
+  assert all_fn.shape == (jax.process_count(), num_thresholds)
+
+  tp = jnp.sum(all_tp, 0)
+  fp = jnp.sum(all_fp, 0)
+  tn = jnp.sum(all_tn, 0)
+  fn = jnp.sum(all_fn, 0)
+
+  return tp, fp, tn, fn
+
+
+class ComputeAUCMetric:
+  """Calculate auc metrics across multiple hosts."""
```

Comment:
We could probably replace all of the custom multihost AUC metric code we wrote with the new `host_all_gather_metrics` function we added, using it in the same way as for ECE. Fine not to do it now, but could be a nice cleanup.

Reply:
Done